### PR TITLE
feat(mastracode): wave duration telemetry — foundation for v1 telemetry system

### PR DESCRIPTION
**Pull request by asibilia**
# Wave Duration Telemetry — Foundation for V1 Telemetry

## What

`workflowState` now emits structured JSONL records at phase/wave boundaries to `.planning/telemetry/<runId>.jsonl`. Foundation for the Wave 1 telemetry program (subagent invocation costs, recall hit/miss rates, review convergence, cross-run aggregator).

## Why

No visibility into:
- How long each phase/wave takes (performance baseline)
- When the pipeline spends time (execution, review iteration cycles)
- How recall filtering impacts pipeline decisions

TelemetryRecord v:1 locked (additive evolution only) so 4 follow-on todos consume it verbatim without coordination.

### New module: `src/state/telemetry.ts`

Exports fail-safe writer + reader:

- `appendTelemetry(kind, meta?, overrides?)` — emits event, never throws
- `readTelemetry(runId)` — parses JSONL, drops malformed lines with warn
- `TelemetryRecord` + v:1 schema (Zod)

### New state field: `PhaseResult.waveStartedAt`

Tracks wave start across:
- `startPhase(new-phase)` — set to now
- `startPhase(RESUME)` — reset to now (currentWave resets to 1)
- `advanceWave()` — update to now

### Hook sites in `workflowState.ts`

Three telemetry events:
1. `start-phase` action → `phase.start` + `wave.start`
2. `advance-wave` action → `wave.end` (with pre-mutation phase/slug/wave in overrides)
3. `complete-phase` action → `wave.end` + `phase.end` (with pre-mutation context)

All hooks fail-safe (try/catch internally).

### Whitelist update

`ROOT_WHITELIST_DIRS` includes `'telemetry'` for straggler-detection allowlist.

## Testing

314/314 tests pass (up from 310 due to test.each split + new invalid-runId test).

- `src/__tests__/telemetry.test.ts` — writer unit tests (append, read, malformed, fail-safe)
- `src/__tests__/luca-store.test.ts` — `waveStartedAt` lifecycle (new, resume, advance)
- `src/__tests__/workflow-state-actions.test.ts` — 3 hook sites emit correct records

## Schema (v:1 locked)

```typescript
{
  v: 1,
  ts: ISO8601,
  runId: string,
  kind: "phase.start" | "wave.start" | "wave.end" | "phase.end",
  phase: string,
  slug: string,
  wave: number,
  complexity: "TRIVIAL" | "SIMPLE" | "MODERATE" | "COMPLEX" | "CRITICAL",
  oversight: "full-auto" | "checkpoint" | "human-in-loop",
  durationMs: number | null,
  meta: Record<string, unknown>
}
```

File: `.planning/telemetry/<runId>.jsonl` (append-only, one record per line).

## Review

- **Iter 1**: 8 items (2 MF + 6 SF) → resolved
- **Iter 2**: 8 items (2 MF + 6 SF) → resolved
- **Iter 3**: Final audit → 0 MF, APPROVED

Security: CWE-117 log injection prevention (sanitized path interpolation).
Simplification: Zombie export removal (isValidRunId audited, 0 callers).
Testing: test.each conversion, explicit edge-case coverage.

## Metrics

Commits: 6 (3 waves + 2 review-fix + 1 changeset)
Tests: 314/314 pass
Files: 5 changed
Changeset: @alecsibilia/luca-mastracode minor bump

## Notes

- Failure modes (ENOSPC, EACCES) logged + swallowed — pipeline never blocked
- File integrity: runId validated /^run_[a-z0-9]+_[a-z0-9]+$/ before path construction
- Lifecycle: waveStartedAt reset on phase resume (currentWave resets to 1)
- Overrides: closing events pass pre-mutation phase/slug/wave to handle state mutation

### CODE REVIEW

File: packages/luca-mastracode/src/tools/workflow-state.ts [hunk 1]
```diff
@@ -757,6 +813,38 @@ export const workflowStateTool = createTool({
...
+                            'phase.end',
+                            {},
+                            {
+                                wave: tPriorWave,
+                                phase: tPriorPhase,
+                                slug: tPriorSlug,
+                                durationMs: tPriorPhaseStartedAt
+                                    ? now -
+                                      new Date(tPriorPhaseStartedAt).getTime()
+                                    : null,
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/workflow-state.ts:878]: Fixed in dfe10a729. Same `finiteOrNull()` guard applied to both `wave.end` and `phase.end` `durationMs` computations in the complete-phase block. Malformed `waveStartedAt` / `startedAt` now yields `durationMs: null` instead of dropping the closing events.

Regression test added: 'complete-phase emits wave.end + phase.end with durationMs:null when timestamps malformed' (seeds both timestamps as garbage strings, asserts both events emit with `durationMs === null`).

File: packages/luca-mastracode/src/state/telemetry.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+        console.warn(
+            `[telemetry] Skipped ${invalidLines.length} invalid ` +
+                `entr${invalidLines.length === 1 ? 'y' : 'ies'} ` +
+                `in ${sanitizeLogMessage(p)} at line${invalidLines.length === 1 ? '' : 's'} ${invalidLines.join(', ')}` +
+                (firstError ? ` (first error: ${firstError})` : '') +
+                '.'
+        )
+    }
+
+    return records
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/state/telemetry.ts]: Fixed in dfe10a729. Wrapped the `readFileSync` + parse loop in a recovery try/catch that returns `[]` with a sanitized `console.warn` (`[telemetry] read failed for <path>: <err>`). The no-throw read contract is now honored even for TOCTOU between `existsSync` and `readFileSync`, EACCES/EPERM, or mid-read filesystem errors.

Regression test added: `telemetry.test.ts` → 'returns [] when readFileSync throws (TOCTOU / EACCES)' (mocks `fs.readFileSync` to throw, asserts `[]` return + warn). 317/317 pass.

File: packages/luca-mastracode/src/tools/workflow-state.ts [hunk 3]
```diff
@@ -636,12 +644,45 @@ export const workflowStateTool = createTool({
...
+                    const priorPhase = preWaveState.currentPhaseName ?? null
+                    const priorSlug = preWaveState.currentPhaseSlug ?? null
+                    const priorWaveNum = preWaveState.currentWave ?? null
+                    const priorEntry = preWaveState.phaseResults?.find(
+                        (r) => r.name === priorPhase
+                    )
+                    const priorWaveStartedAt = priorEntry?.waveStartedAt
+                    const priorDurationMs = priorWaveStartedAt
+                        ? Date.now() - new Date(priorWaveStartedAt).getTime()
+                        : null
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/tools/workflow-state.ts:683]: Fixed in dfe10a729. Added `finiteOrNull()` helper at the top of workflow-state.ts and gated the `priorDurationMs` computation. Non-finite (NaN/Infinity) and negative deltas now emit `durationMs: null`, so the `wave.end` event is still recorded instead of being silently dropped by Zod.

Regression test added: 'advance-wave emits wave.end with durationMs:null when waveStartedAt is malformed' (seeds `waveStartedAt: 'not-an-iso-date'`, asserts `overrides.durationMs === null` while the action still succeeds).

File: packages/luca-mastracode/src/__tests__/telemetry.test.ts [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            throw new Error('ENOSPC: no space left on device')
+        })
+
+        // Must not throw.
+        expect(() => appendTelemetry('phase.start')).not.toThrow()
+        expect(warn).toHaveBeenCalled()
+        const callArg = warn.mock.calls[0]?.[0] ?? ''
+        expect(callArg).toContain('[telemetry] write failed')
+
+        appendSpy.mockRestore()
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/__tests__/telemetry.test.ts:167]: Good catch — fixed in dfe10a729. Switched the production call site to namespace access (`import * as fs from 'node:fs'` → `fs.appendFileSync(...)`) so `spyOn(fs, 'appendFileSync')` in the test now actually intercepts the call. Verified by running the existing ENOSPC test (`does NOT throw when appendFileSync throws (disk full / permission)`) — it now genuinely exercises the failure path rather than silently no-op'ing past it.

Added an explanatory note at the top of telemetry.ts so future contributors don't "clean up" the namespace import back to a named binding.

File: packages/luca-mastracode/src/util/phase-paths.ts [hunk 5]
```diff
@@ -340,3 +340,90 @@ export function RUNS_ROOT(): string {
...
+export function assertValidRunId(runId: unknown): asserts runId is string {
+    if (typeof runId !== 'string' || runId.length === 0) {
+        throw new Error(
+            `Invalid runId: ${JSON.stringify(runId)} (must be non-empty string).`
+        )
+    }
+    if (runId.length > 64) {
+        throw new Error(
+            `Invalid runId: ${JSON.stringify(runId)} (exceeds 64 chars).`
+        )
```
[the remark this answers is not in the document]
Comment (asibilia) [reply packages/luca-mastracode/src/util/phase-paths.ts:436]: Fixed in 39b9c8170. Added a private `displayRunId(runId)` helper in `phase-paths.ts` that returns the verbatim `JSON.stringify` form for strings ≤ 80 chars (preserving debuggability for normal values) and a bounded `"<head40>…<tail8>" (len=N)` summary for longer values. Applied to all four error-message sites in `assertValidRunId` (non-string, length > 64, path-shape, regex). A 10MB tampered `runId` now produces a sub-100-char error message instead of a 10MB one.

Regression test added: `telemetry.test.ts` → 'error message is bounded for very long tampered runId values' (asserts `err.message.length < 500` and contains `len=10000` for a 10_000-char input).

File: .planning/ROADMAP.md [hunk 6]
```diff
@@ -2,22 +2,27 @@
...
 ## Phases
 
-### Phase 1: Phase A — Project preferences foundation
+### Phase 1: Phase 1: Phase A — Project preferences foundation
 - **Status**: complete
 - **Dependencies**: None
 - **WSJF Score**: 1.8
 
-### Phase 2: Phase B — Branching policy refactor
+### Phase 2: Phase 2: Phase B — Branching policy refactor
```
[the remark this answers is not in the document]
Comment (asibilia) [reply .planning/ROADMAP.md]: Fixed in 39b9c8170. Stripped the duplicated `Phase N:` prefix from all 5 phase headings in `.planning/ROADMAP.md` and corrected the dependency name references to match. Headings are now in the canonical `### Phase N: <name>` form, so `manageRoadmap(update-status)` will find them.

Likely root cause: an earlier `manageRoadmap` insert that prepended `Phase N: ` to the user-provided heading name without first stripping any existing prefix — worth a separate guard on the insert path, but out of scope for this PR.
